### PR TITLE
simplify flake

### PR DESCRIPTION
**Pull request by zimbatm**
Basically finishing #6196

/cc @Ericson2314

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]
Thank you!!

**Comment by zimbatm** [comment]
@edolstra any objections to merge?

**Comment by roberth** [comment]
I think this PR should be assigned to a single person who takes full ownership.
If I recall correctly we decided to review together - hence no assignee, but I think this is unnecessary. It is just a refactoring, and anyone can rebase it and then ascertain that either the final attributes remain unchanged, or are perhaps slightly improved somehow.
As this change is sensitive to conflicts, it's better to act decisively rather than go back and forth to let this stall again.
I'd be happy to complete this task.

**Comment by zimbatm** [comment]
The merge conflicts are why I haven't been rebasing it anymore. I was waiting to get a clear signal that this would get merged fairly quickly after. So please tell me "as the Nix team ..." and then I will re-do it.

**Comment by Ericson2314** [comment]
I think we moved it to "in review" in triage without posting a thing here in hopes someone would fix it. I will move back to in discussion so we can get @zimbatm the "clear signal".

**Comment by Ericson2314** [comment]
Or alternatively someone on the @nixos/nix-team maybe feels comfortable approving the concept here async.

**Comment by Ericson2314** [comment]
(sorry hit wrong button!)

**Comment by roberth** [comment]
@zimbatm
Friendly reminder that this has been approved by the Nix team. Would you like to update this PR?

**Comment by zimbatm** [comment]
ok, this is ready for a round of review

File: flake.nix [hunk 1]
```diff
@@ -25,34 +24,36 @@
 
       stdenvs = [ "gccStdenv" "clangStdenv" "clang11Stdenv" "stdenv" "libcxxStdenv" "ccacheStdenv" ];
 
-      forAllSystems = f: nixpkgs.lib.genAttrs systems (system: f system);
-      forAllSystemsAndStdenvs = f: forAllSystems (system:
+      forAllSystems = nixpkgs.lib.genAttrs systems;
+
+      forAllCrossSystems = nixpkgs.lib.genAttrs crossSystems;
+
+      forAllStdenvs = prefix: f:
```
Comment (Ericson2314) [inline flake.nix:31]: @zimbatm and I discussed using `lib.mapAttrs' where a non-`""` prefix is needed instead of adding this argument.

Comment (zimbatm) [reply flake.nix]: fixed!

File: flake.nix [hunk 2]
```diff
@@ -640,67 +598,26 @@
...
+          stdenvs)));
+
+      devShells = let
+        makeShell = pkgs: stdenv:
           with commonDeps { inherit pkgs; };
-          nixpkgsFor.${system}.${stdenv}.mkDerivation {
+          stdenv.mkDerivation {
             name = "nix";
 
+
```
Comment (Ericson2314) [inline flake.nix:620]: blank line that got added.

File: flake.nix [hunk 3]
```diff
@@ -300,24 +304,26 @@
             outputs = [ "out" "dev" "doc" ];
 
             nativeBuildInputs = nativeBuildDeps;
-            buildInputs = buildDeps ++ awsDeps;
+            buildInputs = buildDeps
+              # There were issues building this in the past
```
Comment (Ericson2314) [inline flake.nix:308]: I bet this comment is from me :sweat_smile: but it is kind of cryptic!

Comment (Ericson2314) [reply flake.nix:308]:
```suggestion
              # There have been issues building these dependencies
```
this is a little clearer. I am fine figuring out later if we can cross compile them after all.

**Review by Ericson2314 — APPROVED** [review]
It looks good now! Further cleanups / de-duping can be done in follow-up PRs.

File: flake.nix (hunk 2 above)
Comment (zimbatm) [reply flake.nix]: fixed

**Review by Ericson2314 — APPROVED** [review]

**Review by roberth — APPROVED** [review]
Checked by evaluating some of the attributes. No changes in the default package or alternate stdenv packages. Small improvements to the cross package, such as separate debug info.

```shell
d() { nix-diff "$(nix eval ".?ref=upstream/master#$1" | jq -r .)" "$(nix eval ".#$1" | jq -r .)"; }
```

**Comment by roberth** [comment]
Note to self: have a look at the hydra comparison after this is merged.

**Comment by zimbatm** [comment]
couldn't resist to add two more noop changes :)

**Comment by roberth** [comment]
Lovely! Could you squash the fixups?

**Comment by zimbatm** [comment]
done

**Comment by roberth** [comment]
CI seems to have failed https://github.com/NixOS/nix/actions/runs/4222266488/jobs/7330567806

**Comment by zimbatm** [comment]
caused by f2e427942dbdef460c6ba5641a571388f692bff9

**Comment by zimbatm** [comment]
fixed

**Pull request merged**